modules/calib3d/src/solvepnp.cpp

```diff
@@ -456,4 +398,325 @@ int solveP3P( InputArray _opoints, InputArray _ipoints,
+//    else
+//    {
+//        // use CV_32F if all PnP inputs are CV_32F and outputs are empty
+//        if (_ipoints.depth() == _cameraMatrix.depth() && _ipoints.depth() == _opoints.depth() &&
+//            _rvec.empty() && _tvec.empty())
+//            mtype = _opoints.depth();
+//    }
+
+    Mat cameraMatrix0 = _cameraMatrix.getMat();
+    Mat distCoeffs0 = _distCoeffs.getMat();
+    Mat cameraMatrix = Mat_<double>(cameraMatrix0);
+    Mat distCoeffs = Mat_<double>(distCoeffs0);
+
+    vector<Mat> vec_rvecs, vec_tvecs;
+    if (flags == SOLVEPNP_EPNP || flags == SOLVEPNP_DLS || flags == SOLVEPNP_UPNP)
+    {
+        Mat undistortedPoints;
+        undistortPoints(ipoints, undistortedPoints, cameraMatrix, distCoeffs);
+        epnp PnP(cameraMatrix, opoints, undistortedPoints);
+
+        Mat rvec, tvec, R;
+        PnP.compute_pose(R, tvec);
+        Rodrigues(R, rvec);
+
+        vec_rvecs.push_back(rvec);
+        vec_tvecs.push_back(tvec);
+    }
+    else if (flags == SOLVEPNP_P3P)
+    {
+        CV_Assert(npoints == 4);
+        Mat undistortedPoints;
+        undistortPoints(ipoints, undistortedPoints, cameraMatrix, distCoeffs);
+        p3p P3Psolver(cameraMatrix);
+
+        Mat rvec, tvec, R;
+        bool result = P3Psolver.solve(R, tvec, opoints, undistortedPoints);
+        if (result)
+        {
+            Rodrigues(R, rvec);
+            vec_rvecs.push_back(rvec);
+            vec_tvecs.push_back(tvec);
+        }
+    }
+    else if (flags == SOLVEPNP_AP3P)
+    {
+        CV_Assert(npoints == 4);
+        Mat undistortedPoints;
+        undistortPoints(ipoints, undistortedPoints, cameraMatrix, distCoeffs);
+        ap3p P3Psolver(cameraMatrix);
+
+        Mat rvec, tvec, R;
+        bool result = P3Psolver.solve(R, tvec, opoints, undistortedPoints);
+        if (result)
+        {
+            Rodrigues(R, rvec);
+            vec_rvecs.push_back(rvec);
+            vec_tvecs.push_back(tvec);
+        }
+    }
+    else if (flags == SOLVEPNP_ITERATIVE)
+    {
+        Mat rvec, tvec;
+        if (useExtrinsicGuess)
+        {
+            rvec = _rvec.getMat();
+            tvec = _tvec.getMat();
+        }
+        else
+        {
+            rvec.create(3, 1, mtype);
+            tvec.create(3, 1, mtype);
+        }
+
+        CvMat c_objectPoints = cvMat(opoints), c_imagePoints = cvMat(ipoints);
+        CvMat c_cameraMatrix = cvMat(cameraMatrix), c_distCoeffs = cvMat(distCoeffs);
+        CvMat c_rvec = cvMat(rvec), c_tvec = cvMat(tvec);
+        cvFindExtrinsicCameraParams2(&c_objectPoints, &c_imagePoints, &c_cameraMatrix,
+                                     (c_distCoeffs.rows && c_distCoeffs.cols) ? &c_distCoeffs : 0,
+                                     &c_rvec, &c_tvec, useExtrinsicGuess );
+
+        vec_rvecs.push_back(rvec);
+        vec_tvecs.push_back(tvec);
+    }
+    else if (flags == SOLVEPNP_IPPE)
+    {
+        double val = 0.0;
+        bool planar = isPlanarObjectPoints(opoints, 1e-5, val);
```

Comment:
I have added a method to check that the input points are indeed planar. Otherwise, we return false.

Which threshold to use? Here we have:
https://github.com/opencv/opencv/blob/1f517b8a0293f2574c3dce739e5b273164244cac/modules/calib3d/src/calibration.cpp#L1112

In IPPE there is [#define IPPE_SMALL 1e-7](https://github.com/tobycollins/IPPE/blob/3af0127a30fcb0d35fa2a34f4abd93278e631621/cpp/include/IPPE/ippe.h#L9)

Reply:
The threshold is a bit arbitrary. To keep consistent with OpenCV 1e-3 should be fine.

Reply:
I have changed the threshold to use 1e-3 for IPPE_SMALL too.